CRM/Automateddirectdebit/Job/DirectDebitEvents/PaymentCollectionEvent.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/**
+ * Responsible for emitting direct debit payment collection event (hook).
+ * The dispatched event/hook is called: `automateddirectdebit_PaymentCollectionEvent`
+ * where 3rd party payment processor extensions (such as GoCardless extension) can implement it
+ * to create and collect payments.
+ *
+ */
+class CRM_Automateddirectdebit_Job_DirectDebitEvents_PaymentCollectionEvent {
+
+  public function run() {
+    $pendingInvoicesQuery = $this->buildPendingInvoicesQuery();
+    $pendingInvoices = CRM_Core_DAO::executeQuery($pendingInvoicesQuery->toSQL());
+
+    while ($pendingInvoices->fetch()) {
+      $pendingInvoiceData = $pendingInvoices->toArray();
+      $invoiceTotalPaidAmount = CRM_Core_BAO_FinancialTrxn::getTotalPayments($pendingInvoiceData['contribution_id'], TRUE);
+      if ($invoiceTotalPaidAmount >= $pendingInvoiceData['total_amount']) {
+        continue;
+      }
+
+      $pendingInvoiceData['charge_amount'] = $pendingInvoiceData['total_amount'] - $invoiceTotalPaidAmount;
+
+      $this->markContributionExternalPaymentToBeInProgress($pendingInvoiceData['contribution_id']);
+      $this->dispatchPaymentCollectionEventHook($pendingInvoiceData);
+    }
+  }
+
+  /**
+   * Builds the query to fetch the contributions (invoices)
+   * that match the criteria of invoices that we
+   *
+   * @return CRM_Utils_SQL_Select
+   */
+  public function buildPendingInvoicesQuery() {
+    $recurContributionStatusesToProcess = implode(',', $this->getRecurContributionStatusesIdsToProcess());
+    $mandateActiveStatusId = 1;
```

Comment:
Shouldn't we hardcode the status ID?  I understand that we reserved the active status, but it can be changed.

Reply:
it is marked as "reserved" so the value should never change, but you are right it is better to retrieve it from the db instead of hardcoding it.